algorithm/evocoin.c

```diff
@@ -121,42 +121,37 @@ int nextPerm(uint8_t n[], uint32_t count) {
 }
 
 
-void getAlgoString(char *str, uint32_t count)
-{
-	uint8_t algoList[HASH_FUNC_COUNT];
-	char s[100];
-	char *sptr;
-
+void getAlgoString(uint8_t *algoList, uint32_t count)
+{	
 	initPerm(algoList, HASH_FUNC_COUNT);
-
-	int j;
-
 	int k;
 	for (k = 0; k < count; k++) {
 		nextPerm(algoList, HASH_FUNC_COUNT);
 	}
-
-	sptr = str;
-	for (j = 0; j < HASH_FUNC_COUNT; j++) {
-		if (algoList[j] >= 10)
-			sprintf(sptr, "%c", 'A' + (algoList[j] - 10));
-		else
-			sprintf(sptr, "%u", algoList[j]);
-		sptr++;
-	}
-	*sptr = 0;
 }
 
 
-void evocoin_twisted_code(char *result, const char *ntime, char *code)
+void evocoin_twisted_code(char *result, const char *ntime, uint8_t *code)
 {
 	unsigned char bin[4];
 	uint32_t h32, *be32 = (uint32_t *)bin;
 	hex2bin(bin, ntime, 4);
 	h32 = be32toh(*be32);
 	uint32_t count = getCurrentAlgoSeq(h32, INITIAL_DATE);
 	getAlgoString(code, count);
-	sprintf(result, "_%d_%s_", count, code);
+	char *sptr;
+	int j;
+	int n = sprintf(result, "_%d_", count);
+	sptr = result + n;
+	for (j = 0; j < HASH_FUNC_COUNT; j++) {
```

Comment:
i got idea, a better "converter" based on bin2hex function from util.c

    static const char codes[HASH_FUNC_COUNT] = {'0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'a' };
    for (j = 0; j < HASH_FUNC_COUNT; j++) {
      *s++ = codes[code[j]];
    }

Reply:
I applied this change but now the kernel binary is not saved on disk anymore and it needs to be rebuilded on each new run

Reply:
well, i missed the variable, it need to be *sptr++ = ...

if still wrong, try log the result variable at the end of evocoin_twisted_code. maybe i missed something, today was a long day :/

Reply:
well, the variable name was not an issue, even if I had noticed it the code would not have compiled ;)
I also tried to change 'a' with 'A' before writing to you, with the same result.
I just logged the result variable as you suggested and the weird thing seems that the last character, that should be "_" because of ```strcat(sptr, "_");```, it is instead a semi-random character, sometimes is "_", sometimes is "q", sometimes "1" and sometimes is non-ascii..
You loop seems fine and the post-increment on sptr should work, I can't see why strcat should not work properly anymore after that loop

Reply:
Hi, I was very busy these days, but I found the issue.
The problem was caused by ```strcat```: it searches for the terminating null-character at the end of the string and then concatenates the two strings overwriting the terminating null-character. Now, with the old code the ```sprintf``` added automatically the terminating null-character so strcat worked without issues, but with the new code it needs to be explicitily added or the ```strcat``` behaviour is not defined, that's why the random characters instead than the expected '_'.
I'm going to squash all the commits and push the new code, I initially manually added a terminating null-character at the end of the for loop to let ```strcat``` to do its job, but finally I removed the ```strcat``` operation at all and and added the '_' directly, it seems cleaner to me.
I'm pushing the new code in a couple of minutes